### PR TITLE
Set Default start_date to 3 Years Ago for actions and actionUpdates

### PR DESCRIPTION
**Pull request by rdeshmukh15**
# Description of change

1. According to the impact document at: https://integrations.impact.com/impact-brand/changelog/changes-to-actions-and-action-updates, new changes were implemented on August 5, 2024. One of these changes states that queries for Actions or ActionUpdates cannot have a StartDate that is more than 3 years old.

      As a result of this restriction, customers began encountering errors when the start_date specified in their configuration 
      is older than 3 years. The error message is as follows: 
      ```
    `2024-09-09 15:26:43,856Z tap - ERROR ERROR 400: {"Status":"ERROR","Message":"Parameter 'ActionDateStart' 
       cannot be more than 3 years in the past."}, REASON: Bad Request`
    ```
    
     Hence, as part of fix, added a default start_date to 3 years ago when start_date is more than 3 years old.
2. #34 
3. #35 
4. #28 

# Manual QA steps
 - Tested the changes on customer connection and those are working fine.
 
# Risks
 - low
 
# Rollback steps
 - revert this branch

#### AI generated code
https://internal.qlik.dev/general/ways-of-working/code-reviews/#guidelines-for-ai-generated-code
- [ ] this PR has been written with the help of GitHub Copilot or another generative AI tool

### CODE REVIEW

File: tap_impact/sync.py [hunk 1]
```diff
@@ -143,18 +143,24 @@ def sync_endpoint(client,
...
+        default_date = end_dttm - timedelta(days=3*365)
+        default_date_str = default_date.strftime('%Y-%m-%dT%H:%M:%SZ')
 
-    end_dttm = utils.now()
-    end_dt = end_dttm.date()
-    end_dt_str = end_dt.strftime('%Y-%m-%dT%H:%M:%SZ')
-    start_dttm = end_dttm
-    start_dt = end_dt
+        if stream_name in ('actions', 'action_updates') and last_datetime_dt < default_date:
+            last_datetime = default_date_str
```
Comment (RushiT0122) [inline tap_impact/sync.py:162]: We should log the warning if we are setting start time to default date.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
IMO, we should fail the discovery if date is older than 3 years.

**Comment by rdeshmukh15** [comment]
> IMO, we should fail the discovery if date is older than 3 years.

This rule of having start_date only upto 3 years is for Actions and ActionUpdates.

File: tap_impact/sync.py [hunk 2]
```diff
@@ -143,38 +156,41 @@ def sync_endpoint(client,
...
     else:
         last_datetime = get_bookmark(state, stream_name, start_date)
+        last_datetime_dt = datetime.fromisoformat(last_datetime.replace('Z', '+00:00'))
+        default_date = end_dttm - timedelta(days=3*365)
+        default_date_str = default_date.strftime('%Y-%m-%dT%H:%M:%SZ')
+
+        if stream_name in ('actions', 'action_updates') and last_datetime_dt < default_date:
+            LOGGER.warning('Last datetime {} is older than 3 years, setting to default date {}'.format(last_datetime, default_date_str))
+            last_datetime = default_date_str
+
```
Comment (RushiT0122) [inline tap_impact/sync.py:175]: Can we add a comment for `default_date` and rename it to something more relevant like `min_default_date`?

Also constraint of minimum 3 years date is specific to `actions` and `action_updates`, please mention those stream names in the warning message.

File: tap_impact/sync.py [hunk 3]
```diff
@@ -143,38 +156,41 @@ def sync_endpoint(client,
...
-    date_list = [str(start_dt + timedelta(days=x)) for x in range((end_dt - start_dt).days + 1)]
     endpoint_total = 0
     total_records = 0
     limit = 1000 # PageSize (default for API is 100)
-    for bookmark_date in date_list:
+
+    for start_date_from_ranges, end_date_from_ranges in date_ranges:
+        # Convert dates to strings for API request
+        start_date_str = start_date_from_ranges.strftime('%Y-%m-%dT%H:%M:%SZ')
+        end_date_str = end_date_from_ranges.strftime('%Y-%m-%dT%H:%M:%SZ')
```
Comment (RushiT0122) [inline tap_impact/sync.py:189]:
```suggestion
    for start_date, end_date in date_ranges:
        # Convert dates to strings for API request
        start_date_str = start_date_from_ranges.strftime('%Y-%m-%dT%H:%M:%SZ')
        end_date_str = end_date_from_ranges.strftime('%Y-%m-%dT%H:%M:%SZ')
```

File: tap_impact/sync.py (hunk 2 above)
Comment (RushiT0122) [reply tap_impact/sync.py:175]: Add a unittests for this change for scenarios.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]

File: tap_impact/sync.py (hunk 2 above)
Comment (rdeshmukh15) [reply tap_impact/sync.py]: Added the comment but kept default_date variable name as it is to keep it simpler.
> Can we add a comment for default_date and rename it to something more relevant like min_default_date?

Added

> Also constraint of minimum 3 years date is specific to actions and action_updates, please mention those stream names in the warning message.

File: tests/unittests/test_default_start_date.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+from tap_impact.sync import get_bookmark
+from singer import utils
+
+class TestBookmarkDateHandling(unittest.TestCase):
+
+    @parameterized.expand([
+        # Test case 1: Start date within 3 years
+        ({}, 'actions', '2024-12-01T00:00:00Z', '2024-12-01T00:00:00Z'),
+        # Test case 2: Bookmark date within 3 years but start date is older
+        ({"bookmarks": {"actions": "2024-12-01T00:00:00Z"}}, 'actions', '2014-01-01T00:00:00Z', '2024-12-01T00:00:00Z'),
```
Comment (RushiT0122) [inline tests/unittests/test_default_start_date.py:13]: Parameters should be be relative to current date so that test will be durable.

**Review by RushiT0122 — APPROVED** [review]

File: tap_impact/sync.py [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
-from singer.utils import strptime_to_utc, strftime
+from singer.utils import strptime_to_utc
 from tap_impact.transform import transform_json
 from tap_impact.streams import STREAMS
 
 LOGGER = singer.get_logger()
 BASE_URL = 'https://api.impact.com'
+# Window size implemented for actions and actionUpdates streams
+# as startDate and EndDate can just be 45 days apart
+DEFAULT_WINDOW_SIZE = 45
```
Comment (somethingmorerelevant) [inline tap_impact/sync.py:13]: any specific reason to keep it 45 days? rather than a standard 1 month or 30 days date window?

Comment (rdeshmukh15) [reply tap_impact/sync.py:13]: This is the PR referenced in description which tells about this in detail: https://github.com/singer-io/tap-impact/pull/35

**Pull request merged**